source/adapters/level_zero/queue.hpp

```diff
@@ -687,8 +699,38 @@ struct ur_queue_handle_t_ : _ur_object {
 
   // Threshold for cleaning up the EventList for immediate command lists.
   size_t getImmdCmmdListsEventCleanupThreshold();
+
+  // Pointer to the unified handle.
+  ur_queue_handle_t_ *UnifiedHandle;
 };
 
+// Unified handle that represents either legacy Queue or new dispatcher.
+struct ur_queue_handle_t_ {
+  template <typename QType, typename... Args>
+  ur_queue_handle_t_(std::in_place_type_t<QType> tag, Args &&...args)
+      : Queue(tag, std::forward<Args>(args)...) {
+    if constexpr (std::is_same_v<QType, ur_queue_handle_legacy_t_>) {
+      std::get<ur_queue_handle_legacy_t_>(Queue).UnifiedHandle = this;
+    }
+  }
+
+  std::variant<ur_queue_handle_legacy_t_, v2::ur_queue_dispatcher_t> Queue;
```

Comment:
do we need the intermediate type `v2::ur_queue_dispatcher_t` ? Wouldn't this be simpler (and with one less cache-miss) if we just directly stored the appropriate queue types?

Reply:
I assume that for the new path, we will want to create a single API (enqueue(), synchronize(), etc.) that would be implemented by the dispatcher. When, for example, enqueue() is called on a dispatcher, it would just forward the call to the specific implementation (a specific alternative in the variant).

If we make the legacy path part of the dispatcher, we would need to implement those functions for the legacy path as well (I don't think that makes much sense since the legacy queue does not have a very well-defined interface) or special-case the legacy path. I think the approach with 2 level variant just makes it a bit more explicit and clear (but there's no significant difference otherwise IMO).

Reply:
> I assume that for the new path, we will want to create a single API (enqueue(), synchronize(), etc.) that would be implemented by the dispatcher. When, for example, enqueue() is called on a dispatcher, it would just forward the call to the specific implementation (a specific alternative in the variant).

I was thinking that a dispatcher would just have all the functions like urEnqueueKernelLaunch, and call them directly for each API entry point. Things like enqueue() would be just some shared helper between all 'new' implementations you could use to construct an enqueue function.

> I think the approach with 2 level variant just makes it a bit more explicit and clear (but there's no significant difference otherwise IMO).

Now that I understand the intention what's meant to be inside a dispatcher (I thought you wanted it to be a pointer to some object) I agree that the difference is inconsequential.

Reply:
> I was thinking that a dispatcher would just have all the functions like urEnqueueKernelLaunch, and call them directly for each API entry point.

I see, yes, that would also be possible although probably a bit more verbose - we would need to define all those functions for all queue alternatives. 

Also, I think there are some interactions with different entities (kernel, event, etc.). Right now, those other entities just use variables from the queue directly so we would probably need to abstract this somehow.